### DISCUSSION TITLE
How to install oqs in strongSwan

### DISCUSSION THREAD

**Discussion by ghost** [strongswan/strongswan#1031] [Q&A] [answered]
I want to install the oqs library to use PQC in strongSwan.
However, when installing the oqs library, I get the following error:

> depbase=echo oqs_kem.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||';\
/bin/bash ../../../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../../../..  -I../../../../src/libstrongswan  -rdynamic -g -O2 -Wall -Wno-format -Wno-format-security -Wno-pointer-sign -include /home/mskim/src/IPSec/strongswan-6.0dr15/config.h -MT oqs_kem.lo -MD -MP -MF $depbase.Tpo -c -o oqs_kem.lo oqs_kem.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../../.. -I../../../../src/libstrongswan -rdynamic -g -O2 -Wall -Wno-format -Wno-format-security -Wno-pointer-sign -include /home/mskim/src/IPSec/strongswan-6.0dr15/config.h -MT oqs_kem.lo -MD -MP -MF .deps/oqs_kem.Tpo -c oqs_kem.c  -fPIC -DPIC -o .libs/oqs_kem.o
oqs_kem.c:24:10: fatal error: oqs/oqs.h: No such file or directory
   24 | #include <oqs/oqs.h>
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Makefile:616: oqs_kem.lo] Error 1

How can I fix that error?

Here's the workaround I've applied:

- Download liboqs (https://github.com/open-quantum-safe/liboqs)
- Make oqs.h file in strongSwan's oqs path
- Copy liboqs files to strongSwan's oqs

However, all of the above methods failed.

**Comment by tobiasbrunner** [comment, marked as answer]
You need to install _liboqs_ on your system (follow [the instructions here](https://github.com/open-quantum-safe/liboqs#linuxmacOS)) so the library and header file are found when strongSwan's configure script is run with `--enable-oqs`.

**Comment by ghost** [comment]
Thanks for writing the answer.

However, as a result of doing as you suggested, I get the following error:

> configure: error: Open Quantum-Safe library oqs not found

Do I need to modify environment variables after installing liboqs ?

**Reply by tobiasbrunner** [reply]
That should not be necessary. But you may have to run `ldconfig` (depends on where the library was installed).

**Reply by AlothNaveen** [reply]
@m1nxo @tobiasbrunner  @glguy @sorah @rfrancoise  
 How do I enable OQS from the Android.mk file of the StrongSwan Android client, and how do I specify the include and library paths in Android.mk? Are there any changes needed on the library side to include OQS in the Android client?

**Reply by AlothNaveen** [reply]
[arm64-v8a] SharedLibrary  : libstrongswan.so
  
  C++ build system [build] failed while executing:
      /home/naveen/Android/Sdk/ndk/26.1.10909125/ndk-build \
        NDK_PROJECT_PATH=null \
        APP_BUILD_SCRIPT=/home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/src/main/jni/Android.mk \
        NDK_APPLICATION_MK=/home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/src/main/jni/Application.mk \
        APP_ABI=arm64-v8a \
        NDK_ALL_ABIS=arm64-v8a \
        NDK_DEBUG=1 \
        APP_PLATFORM=android-21 \
        NDK_OUT=/home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/build/intermediates/cxx/Debug/6nb676q2/obj \
        NDK_LIBS_OUT=/home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/build/intermediates/cxx/Debug/6nb676q2/lib \
        -j16 \
        androidbridge \
        charon \
        imcv \
        ipsec \
        strongswan \
        tnccs \
        tncif \
        tpmtss
    from /home/naveen/strongswan-6.0.0beta6/src/frontends/android/app
  /home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/src/main/jni/../../../../../../..//src/libstrongswan/crypto/key_exchange.c:529:50: warning: format specifies type 'char *' but the argument has type 'char (*)[3]' [-Wformat]
          if (!algname || sscanf(algname, "%2s%1u_%255s", &prefix, &num, alg) != 3 ||
                                           ~~~            ^~~~~~~
  1 warning generated.
  ld.lld: error: undefined symbol: OQS_randombytes_switch_algorithm
  >>> referenced by oqs_kem.c:284 (src/main/jni/../../../../../../../src/libstrongswan/plugins/oqs/oqs_kem.c:284)
  >>>               /home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/build/intermediates/cxx/Debug/6nb676q2/obj/local/arm64-v8a/objs-debug/strongswan/plugins/oqs/oqs_kem.o:(oqs_kem_create)
  >>> referenced by oqs_private_key.c:236 (src/main/jni/../../../../../../../src/libstrongswan/plugins/oqs/oqs_private_key.c:236)
  >>>               /home/naveen/strongswan-6.0.0beta6/src/frontends/android/app/build/intermediates/cxx/Debug/6nb676q2/obj/local/arm64-v8a/objs-debug/strongswan/plugins/oqs/oqs_private_key.o:(oqs_private_key_create_empty)
![Screenshot from 2024-07-10 11-27-35](https://github.com/strongswan/strongswan/assets/64768601/59fc69a4-4b91-4681-9e7d-79013f4b5801)
![Screenshot from 2024-07-10 11-27-29](https://github.com/strongswan/strongswan/assets/64768601/57aa534e-1b14-4580-97c9-4c3e33670e66)
![Screenshot from 2024-07-10 11-27-18](https://github.com/strongswan/strongswan/assets/64768601/599d7346-e913-4f32-abc2-209986c67a44)
